File: Assets/Typescript~/src/Server/Services/Match/BW/BWSpawnService.ts
import { CoreServerSignals } from "@Easy/Core/Server/CoreServerSignals";
import { DenyRegionService } from "@Easy/Core/Server/Services/Block/DenyRegionService";
import { EntityService } from "@Easy/Core/Server/Services/Entity/EntityService";
import { PlayerService } from "@Easy/Core/Server/Services/Player/PlayerService";
import { TeamService } from "@Easy/Core/Server/Services/Team/TeamService";
import { CharacterEntity } from "@Easy/Core/Shared/Entity/Character/CharacterEntity";
import { EntityPrefabType } from "@Easy/Core/Shared/Entity/EntityPrefabType";
import { Player } from "@Easy/Core/Shared/Player/Player";
import { MathUtil } from "@Easy/Core/Shared/Util/MathUtil";
import { SignalPriority } from "@Easy/Core/Shared/Util/Signal";
import { Task } from "@Easy/Core/Shared/Util/Task";
import { TimeUtil } from "@Easy/Core/Shared/Util/TimeUtil";
import { Dependency, OnStart, Service } from "@easy-games/flamework-core";
import { ServerSignals } from "Server/ServerSignals";
import { MatchState } from "Shared/Match/MatchState";
import { BedService } from "../BedService";
import { LoadedMap } from "../Map/LoadedMap";
import { MapService } from "../Map/MapService";
import { MatchService } from "../MatchService";
import { BWService } from "./BWService";

/** Spawn delay on join in seconds. */
const SPAWN_DELAY_ON_JOIN = 0;
/** Spawn platform height offset. */
const SPAWN_PLATFORM_HEIGHT_OFFSET = new Vector3(0, 0, 0);

@Service({})
export class BWSpawnService implements OnStart {
	/** Loaded map data. */
	private loadedMap: LoadedMap | undefined;

	constructor(
		private readonly bwService: BWService,
		private readonly playerService: PlayerService,
		private readonly mapService: MapService,
		private readonly matchService: MatchService,
		private readonly entityService: EntityService,
		private readonly bedService: BedService,
	) {
		ServerSignals.MapLoad.Connect((event) => {
			const position = event.LoadedMap.GetSpawnPlatform();
			const entity = this.entityService.SpawnEntityForPlayer(
				undefined,
				EntityPrefabType.HUMAN,
				position.Position.add(new Vector3(-3, 2, 3)),
			);
			entity.AddHealthbar();
		});

		ServerSignals.MapLoad.Connect((event) => {
			for (const team of Dependency<TeamService>().GetTeams()) {
				const spawnPos = this.mapService.GetLoadedMap()?.GetWorldPosition(team.id + "_spawn");
				if (spawnPos) {
					Dependency<DenyRegionService>().CreateDenyRegion(
						MathUtil.FloorVec(spawnPos.Position),
						new Vector3(3, 3, 3),
					);
				}
			}
		});
	}

	OnStart(): void {
		Task.Spawn(() => {
			this.loadedMap = this.mapService.WaitForMapLoaded();
			/* Spawn entity on join. */
			CoreServerSignals.PlayerJoin.ConnectWithPriority(SignalPriority.MONITOR, (event) => {
				Task.Delay(SPAWN_DELAY_ON_JOIN, () => {
					const team = event.player.GetTeam();
					print("join team: " + team?.id);
					if (this.matchService.GetState() > MatchState.PRE && team && this.bedService.IsBedDestroyed(team)) {
						this.bwService.EliminatePlayer(event.player);
						return;
					}
					this.SpawnPlayer(event.player);
				});
			});

			CoreServerSignals.EntityDeath.ConnectWithPriority(SignalPriority.LOW, (event) => {
				if (!this.matchService.IsRunning()) return;

				const maxTime = 20 * 60;
				const baseRespawnTime = 3;
				const maxRespawnTime = 7;

				const currentTime = TimeUtil.GetServerTime() - this.matchService.GetMatchStartTime();
				const ratio = math.min(currentTime / maxTime, 1);
				const respawnTime = baseRespawnTime + (maxRespawnTime - baseRespawnTime) * ratio;

				event.respawnTime = respawnTime;
			});

			/* Listen for entity death, respawn if applicable. */
			CoreServerSignals.EntityDeath.ConnectWithPriority(SignalPriority.MONITOR, (event) => {
				if (!this.matchService.IsRunning()) return;
				if (event.entity instanceof CharacterEntity && !this.bwService.winnerDeclared) {
					Task.Delay(event.respawnTime, () => {
						if (event.entity.player && !this.bwService.IsPlayerEliminated(event.entity.player)) {
							this.SpawnPlayer(event.entity.player);
						}
					});
				}
			});

			/* Listen for match start and teleport players. */
			ServerSignals.MatchStart.Connect(() => {
				this.playerService.GetPlayers().forEach((player) => {
					this.TeleportPlayerToSpawn(player);
				});
			});
		});

		CoreServerSignals.BeforeEntitySpawn.Connect((event) => {
			if (this.matchService.IsRunning() && event.player) {
				const team = event.player.GetTeam();
				if (!team) return;
				const teamSpawnPosition = this.mapService.GetLoadedMap()?.GetWorldPosition(team.id + "_spawn");
				if (teamSpawnPosition) {
					const pos = teamSpawnPosition.Position.add(new Vector3(0, 0.2, 0));
					event.spawnPosition = pos;
				}
			}
		});
	}

	public SpawnPlayer(player: Player): void {
		const entity = Dependency<EntityService>().SpawnEntityForPlayer(player, EntityPrefabType.HUMAN);
		entity.AddHealthbar();
	}

	/** Teleports player to match spawn location on match start. */
	public TeleportPlayerToSpawn(player: Player): void {
		/* Teleport to team spawn location. */
		const team = player.GetTeam();
		if (!team) return;
		const teamSpawnPosition = this.mapService.GetLoadedMap()?.GetWorldPosition(team.id + "_spawn");
		if (teamSpawnPosition) {
			const pos = teamSpawnPosition.Position.add(new Vector3(0, 0.2, 0));
			player.character?.Teleport(pos, teamSpawnPosition.Rotation.eulerAngles);
		}
	}
}

File: Assets/Bundles/Server/Resources/TS/Services/Match/BW/BWSpawnService.lua
-- Compiled with unity-ts v2.1.0-75
local TS = require("Shared/include/RuntimeLib")
local _flamework_core = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init")
local Reflect = _flamework_core.Reflect
local Flamework = _flamework_core.Flamework
local Service = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Service
local ServerSignals = require("Server/TS/ServerSignals").ServerSignals
local CharacterEntity = require("Shared/TS/Entity/Character/CharacterEntity").CharacterEntity
local EntityPrefabType = require("Shared/TS/Entity/EntityPrefabType").EntityPrefabType
local ItemStack = require("Shared/TS/Inventory/ItemStack").ItemStack
local ItemType = require("Shared/TS/Item/ItemType").ItemType
local Task = require("Shared/TS/Util/Task").Task
-- * Spawn delay on join in seconds.
local SPAWN_DELAY_ON_JOIN = 1
-- * Spawn platform height offset.
local SPAWN_PLATFORM_HEIGHT_OFFSET = Vector3.new(0, 0, 0)
local BWSpawnService
do
	BWSpawnService = setmetatable({}, {
		__tostring = function()
			return "BWSpawnService"
		end,
	})
	BWSpawnService.__index = BWSpawnService
	function BWSpawnService.new(...)
		local self = setmetatable({}, BWSpawnService)
		return self:constructor(...) or self
	end
	function BWSpawnService:constructor(bwService, playerService, mapService, matchService)
		self.bwService = bwService
		self.playerService = playerService
		self.mapService = mapService
		self.matchService = matchService
	end
	function BWSpawnService:OnStart()
		Task:Spawn(function()
			self.loadedMap = self.mapService:WaitForMapLoaded()
			-- Spawn entity on join.
			ServerSignals.PlayerJoin:connect(function(event)
				Task:Delay(SPAWN_DELAY_ON_JOIN, function()
					return (Flamework.resolveDependency("Bundles/Server/Services/Global/Entity/EntityService@EntityService")):SpawnEntityForPlayer(event.player, EntityPrefabType.HUMAN)
				end)
			end)
			-- Listen for entity death, respawn if applicable.
			ServerSignals.EntityDeath:Connect(function(event)
				if not self.matchService:IsRunning() then
					return nil
				end
				if TS.instanceof(event.entity, CharacterEntity) and not self.bwService.winnerDeclared then
					Task:Delay(event.respawnTime, function()
						if event.entity.player and not self.bwService:IsPlayerEliminated(event.entity.player) then
							(Flamework.resolveDependency("Bundles/Server/Services/Global/Entity/EntityService@EntityService")):SpawnEntityForPlayer(event.entity.player, EntityPrefabType.HUMAN)
						end
					end)
				end
			end)
			-- Listen for entity spawn and give starter inventory.
			ServerSignals.EntitySpawn:Connect(function(event)
				if TS.instanceof(event.Entity, CharacterEntity) then
					self:giveStarterInventory(event.Entity)
				end
			end)
			-- Listen for match start and teleport players.
			ServerSignals.MatchStart:connect(function()
				local _exp = self.playerService:GetPlayers()
				local _arg0 = function(player)
					self:TeleportPlayerOnMatchStart(player)
				end
				for _k, _v in _exp do
					_arg0(_v, _k - 1, _exp)
				end
			end)
		end)
		ServerSignals.BeforeEntitySpawn:connect(function(event)
			if self.matchService:IsRunning() and event.player then
				local team = event.player:GetTeam()
				if not team then
					return nil
				end
				local teamSpawnPos = self.loadedMap:GetSpawnPositionForTeam(team)
				if not teamSpawnPos then
					return nil
				end
				local _position = teamSpawnPos.Position
				local _vector3 = Vector3.new(0, 0.2, 0)
				event.spawnPosition = _position + _vector3
			end
		end)
	end
	function BWSpawnService:TeleportPlayerOnMatchStart(player)
		-- Teleport to team spawn location.
		local team = player:GetTeam()
		if not team then
			return nil
		end
		local teamSpawnPos = self.loadedMap:GetSpawnPositionForTeam(team)
		if not teamSpawnPos then
			return nil
		end
		local _result = player.Character
		if _result ~= nil then
			_result = _result.gameObject:GetComponent("EntityDriver")
		end
		local humanoid = _result
		if humanoid then
			local _fn = humanoid
			local _position = teamSpawnPos.Position
			local _vector3 = Vector3.new(0, 0.2, 0)
			_fn:Teleport(_position + _vector3)
		end
	end
	function BWSpawnService:giveStarterInventory(entity)
		local inv = entity:GetInventory()
		inv:SetItem(0, ItemStack.new(ItemType.STONE_SWORD, 1))
		inv:SetItem(1, ItemStack.new(ItemType.STONE_PICKAXE, 1))
		-- inv.SetItem(2, new ItemStack(ItemType.WHITE_WOOL, 100));
		inv:SetItem(2, ItemStack.new(ItemType.STONE, 100))
		inv:AddItem(ItemStack.new(ItemType.WOOD_BOW, 1))
		inv:AddItem(ItemStack.new(ItemType.WOOD_ARROW, 100))
		inv:AddItem(ItemStack.new(ItemType.GRASS, 100))
		inv:AddItem(ItemStack.new(ItemType.TELEPEARL, 100))
		-- inv.SetItem(4, new ItemStack(ItemType.WOOD_BOW, 1));
		-- inv.SetItem(5, new ItemStack(ItemType.TELEPEARL, 100));
		-- inv.SetItem(6, new ItemStack(ItemType.FIREBALL, 100));
		-- inv.SetItem(7, new ItemStack(ItemType.WOOD_ARROW, 100));
		-- inv.AddItem(new ItemStack(ItemType.TALL_GRASS, 100));
	end
end
-- (Flamework) BWSpawnService metadata
Reflect.defineMetadata(BWSpawnService, "identifier", "Bundles/Server/Services/Match/BW/BWSpawnService@BWSpawnService")
Reflect.defineMetadata(BWSpawnService, "flamework:parameters", { "Bundles/Server/Services/Match/BW/BWService@BWService", "Bundles/Server/Services/Global/Player/PlayerService@PlayerService", "Bundles/Server/Services/Match/Map/MapService@MapService", "Bundles/Server/Services/Match/MatchService@MatchService" })
Reflect.defineMetadata(BWSpawnService, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(BWSpawnService, "$:flamework@Service", Service, { {} })
return {
	BWSpawnService = BWSpawnService,
}
-- ----------------------------------
-- ----------------------------------
